File: service_backend/app/schemas/app_store.py
"""App Store wire shapes (plan 08 §7) - camelCase out, matching types/app-store.ts."""
from datetime import datetime
from typing import List, Optional


from app.schemas.base import ApiModel

from app.models.module import Module, TenantModule, parse_version


class StoreModuleOut(ApiModel):
    name: str
    title: str
    description: str
    icon: Optional[str] = None
    version: str
    status: Optional[str] = None
    installedVersion: Optional[str] = None
    updateAvailable: bool = False
    installedAt: Optional[datetime] = None
    # Module platform v2 (plan sprint-3/10): declared deps + provided
    # capabilities + the errored/availability flags for dependency-aware UX.
    requires: List[dict] = []
    optional: List[dict] = []
    provides: List[dict] = []
    errored: bool = False
    errorMessage: Optional[str] = None
    # True when this module's hard requires are met for the tenant (else the
    # install button warns + offers cascade-with-consent).
    availabilityOk: bool = True


class InstalledModuleOut(ApiModel):
    module: str
    status: str
    version: str


class UninstallRequest(ApiModel):
    confirmName: str


def store_module_out(
    module: Module,
    state: Optional[TenantModule],
    *,
    manifest: Optional[dict] = None,
    errored: Optional[str] = None,
    availability_ok: bool = True,
) -> StoreModuleOut:
    update_available = bool(
        state and parse_version(module.version) > parse_version(state.installed_version)
    )
    manifest = manifest or {}
    return StoreModuleOut(
        name=module.name,
        title=module.title,
        description=module.description or "",
        icon=module.icon,
        version=module.version,
        status=state.status if state else None,
        installedVersion=state.installed_version if state else None,
        updateAvailable=update_available,
        installedAt=state.installed_at if state else None,
        requires=manifest.get("requires", []) or [],
        optional=manifest.get("optional", []) or [],
        provides=manifest.get("provides", []) or [],
        errored=errored is not None,
        errorMessage=errored,
        availabilityOk=availability_ok,
    )

File: service_frontend/types/app-store.ts
/**
 * App Store types (plan 08) - per-tenant module lifecycle. The catalog is
 * global (synced from on-disk manifests); install state is per tenant
 * (`tenant_modules`). Code is global - "update" provisions the tenant's DATA
 * to the current code version (D3/D4: version gating, never code pinning).
 */

/** Per-tenant install status - code branches only on these. */
export type ModuleStatus = 'ACTIVE' | 'INACTIVE';

/** Lifecycle actions sharing one POST shape (uninstall is separate - typed confirm). */
export type StoreAction = 'install' | 'deactivate' | 'reactivate' | 'update';

/** Catalog entry + the viewing tenant's install state (GET /app-store/modules). */
export interface StoreModule {
  /** manifest `module_name` - the API identity. */
  name: string;
  title: string;
  description: string;
  icon: string | null;
  /** Current CODE version (manifest) - global for all tenants. */
  version: string;
  /** null = not installed for this tenant. */
  status: ModuleStatus | null;
  /** What this tenant is provisioned at - gates features (D4). */
  installedVersion: string | null;
  /** True when `version` > `installedVersion` (server-computed). */
  updateAvailable: boolean;
  installedAt: string | null; // ISO
  /** Module platform v2 (plan sprint-3/10) - declared deps + capabilities. */
  requires?: { name: string; version?: string }[];
  optional?: { name: string; version?: string }[];
  provides?: { capability: string; version: number }[];
  /** A module that failed to load at boot (self-disabled; not installable). */
  errored?: boolean;
  errorMessage?: string | null;
  /** False when this module's hard requires aren't met for the tenant. */
  availabilityOk?: boolean;
}

/** Lightweight row for menu gating (GET /app-store/installed). */
export interface InstalledModule {
  module: string;
  status: ModuleStatus;
  version: string;
}

/** Display state for the storefront StatusBadge (derived, not stored). */
export type StoreModuleBadge = 'NOT_INSTALLED' | 'ACTIVE' | 'INACTIVE' | 'UPDATE_AVAILABLE';

export function moduleBadge(m: StoreModule): StoreModuleBadge {
  if (m.status === null) return 'NOT_INSTALLED';
  if (m.status === 'INACTIVE') return 'INACTIVE';
  return m.updateAvailable ? 'UPDATE_AVAILABLE' : 'ACTIVE';
}
